eyelid-match-ops/src/encrypted.rs

```diff
@@ -1 +1,210 @@
-//! Iris matching operations on fully homomorphic encrypted, polynomial-encoded bit vectors.
+//! Iris matching operations on homomorphic encrypted, polynomial-encoded bit vectors.
+
+use itertools::Itertools;
+use num_bigint::{BigInt, BigUint};
+use rand::rngs::ThreadRng;
+
+use crate::iris::conf::IrisConf;
+use crate::primitives::poly::Poly;
+use crate::{
+    encoded::{MatchError, PolyCode, PolyQuery},
+    primitives::yashe::{Ciphertext, Message, PrivateKey, PublicKey, Yashe},
+    EncodeConf, PolyConf, YasheConf,
+};
+
+pub mod test;
+
+/// An encrypted iris code, encoded in polynomials. To be stored in the database.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct EncryptedPolyCode<C: EncodeConf>
+where
+    C::PlainConf: YasheConf,
+    <C::PlainConf as PolyConf>::Coeff: From<u128> + From<u64> + From<i64>,
+{
+    /// The encrypted polynomials, encoding data, one block of rows each. Storage variant.
+    data: Vec<Ciphertext<C::PlainConf>>,
+    /// The encrypted mask polynomials.
+    masks: Vec<Ciphertext<C::PlainConf>>,
+}
+
+/// An encrypted iris code, encoded in polynomials. To be matched against EncryptedPolyCode.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct EncryptedPolyQuery<C: EncodeConf>
+where
+    C::PlainConf: YasheConf,
+    <C::PlainConf as PolyConf>::Coeff: From<u128> + From<u64> + From<i64>,
+{
+    /// The encrypted polynomials, encoding data, one block of rows each. Query variant.
+    data: Vec<Ciphertext<C::PlainConf>>,
+    /// The encrypted mask polynomials.
+    masks: Vec<Ciphertext<C::PlainConf>>,
+}
+
+/// -1 is encoded as Q-1, so we need to convert it to work modulo T.
+/// Given a vector of polynomials, for each coefficient, if it is larger than Q-1/2 then add T.
+/// Otherwise do nothing.
+pub fn convert_negative_coefficients<C: EncodeConf>(polys: &mut [Poly<C::PlainConf>])
+where
+    <C as EncodeConf>::PlainConf: YasheConf,
+    <<C as EncodeConf>::PlainConf as PolyConf>::Coeff: From<i64>,
+{
+    #[allow(unused_mut)]
+    for mut poly in polys {
+        Poly::coeffs_modify_non_zero(poly, |coeff: &mut <C::PlainConf as PolyConf>::Coeff| {
+            let mut coeff_res = C::PlainConf::coeff_as_big_int(*coeff);
```

Comment:
We don't need to make this change now, but I wanted to write it down when I thought of it.

```suggestion
            // TODO: benchmark comparing `Coeff`s and putting `coeff_res` inside the `if`, it should be faster
            let mut coeff_res = C::PlainConf::coeff_as_big_int(*coeff);
```